File: impls/prolog/step1_read_print.pl
% -*- mode: prolog; -*- select prolog mode in the emacs text editor

:- initialization(main, main).

:- consult([printer, reader, types, utils]).

% Read

mal_read(Ast) :-
    write("user> "),
    read_line_to_string(current_input, Line),
    (Line = end_of_file -> throw(exit_repl) ; true),
    (rl_add_history(Line) -> true ; true), % fails for duplicate lines
    read_str(Line, Ast).

% Eval

eval(Ast, Ast).

% Print

print(Ast) :- format("~F\n", [Ast]).

% REP

rep :-
    mal_read(Ast),
    eval(Ast, Evaluated),
    print(Evaluated).

% Main program

repl :-
    catch(rep, mal_error(Message), writeln(Message)),
    repl.

main(_Argv) :-
    getenv("HOME", Home),
    string_concat(Home, "/.mal-history", History),
    (exists_file(History) -> rl_read_history(History) ; true),

    catch(repl, exit_repl, nl),

    (rl_write_history(History) -> true ; true).

File: impls/lua.2/step1_read_print.lua
Reader = require "reader"
Printer = require "printer"

types = require "types"
--local throw = types.throw
local Err = types.Err
local is_instanceOf = types.isinstanceof

function raw_read(prompt)
  io.write(prompt)
  local v = io.read()
  if v == nil then
    io.write('\n')
  end
  return v
end


function READ(v)
  return  Reader.read_str(v)
end

function EVAL(a)
  return a
end

function PRINT(a)
  print(Printer.stringfy_val(a, true))
end

function rep(str)
  return PRINT(EVAL(READ(str)))
end



function main()
    local line = ''
    while true do
      local line = raw_read('user> ')
      if line == nil then
        break
      end
      status, err = pcall( function () print(rep(line)) end)
      if not status then
        if is_instanceOf(err, Err) then
          err = Printer.stringfy_val(err)
        end
        print(err)
        print(debug.traceback())
      end
    end
end

main()
